mne/decoding/gsoc.py

```diff
@@ -0,0 +1,523 @@
+from .mixin import TransformerMixin
+from ..epochs import _BaseEpochs
+from ..import Covariance, pick_types
+from ..cov import _regularized_covariance
+from ..preprocessing import Xdawn
+
+
+class _EpochsTransformerMixin(TransformerMixin):
+    """Mixin class for reshaping data to Epoch's standard shape
+
+    This class is meant to be inherited by transformers that are to be
+    used in scikit-learn pipeline. It provides functionality to convert
+    data matrix into 3D.
+
+    Parameters
+    ----------
+    n_chan : int (default : None)
+        The number of channels. Used for reshaping data matrix into 3D.
+        If none the matrix should be 3D else error is raised.
+    """
+
+    def __init__(self, n_chan=None):
+        self.n_chan = n_chan
+        self._check_init()
+
+    def fit(self, X, y=None):
+        """No use here. Added for scikit-learn compatibility.
+
+        Parameters
+        ----------
+        X : numpy array, shape(n_epochs, n_chans, n_times) or
+                              (n_epochs, n_chans * n_times)
+            The data to be reshaped into 3D. `n_chan` is used in 3D or 4D
+            matrix.
+        y : None
+            Used for scikit-learn compatibility
+
+        Returns
+        -------
+        self : Instance of EpochsTransformerMixin
+            Return the same object.
+        """
+        return self
+
+    def transform(self, X):
+        """No use here. Added for scikit-learn compatibility.
+
+        Parameters
+        ----------
+        X : numpy array, shape(n_epochs, n_chans, n_times) or
+                              (n_epochs, n_chans * n_times)
+            The data to be reshaped into 3D. `n_chan` is used in 3D or 4D
+            matrix.
+
+        Returns
+        -------
+        X : numpy ndarray
+            The same array.
+        """
+        return X
+
+    def _reshape(self, X):
+        """Recontruct epochs to get a n_trials * n_chans * n_times.
+
+        Parameters
+        ----------
+        X : numpy array, shape(n_epochs, n_chans, n_times) or
+                              (n_epochs, n_chans * n_times)
+            The data to be reshaped into 3D. `n_chan` is used in 3D or 4D
+            matrix.
+
+        Returns
+        -------
+        X : numpy ndarray of shape (n_trials, n_chans, n_times)
+            Transformed data.
+        """
+        if isinstance(X, _BaseEpochs):
+            picks = pick_types(X.info, meg=True, eeg=True, seeg=True,
+                               ecog=True)
+            X = X._data[picks]
+        elif not isinstance(X, np.ndarray):
+            raise ValueError('X must be an Epochs or a 2D or 3D array, got '
+                             '%s instead' % type(X))
+        elif (X.ndim != 3) and (self.n_chan is None):
+            raise ValueError("n_chan must be provided to convert it to 3D")
+        elif (X.ndim != 3) and (self.n_chan is not None):
+            n_epoch = len(X)
+            n_time = np.prod(X.shape[1:]) // self.n_chan
```

Comment:
yeah I know that, but I fear this will give unexpected results if it's not exactly divisible and the error will pass silently